server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -390,17 +391,19 @@ private void executeRequest(
             System::nanoTime
         );
         SearchRequest searchRequest;
+        ActionListener<SearchResponse> listener;
         try {
-            searchRequest = searchPipelineService.transformRequest(originalSearchRequest);
+            Pipeline searchPipeline = searchPipelineService.resolvePipeline(originalSearchRequest);
```

Comment:
Hm ... the process becomes 2 step now, to be fair I would expect the `searchPipelineService` to take care of pipeline and request transformation. What do you think if we only have one step action:

```
class  SearchPipelineService {
    PipelinedSearchRequest apply(SearchRequest request) {
      // resolve pipeline
      // transform the request
    }
}
```

Which returns:

```
    class PipelinedSearchRequest {
        private final Pipeline pipeline;
        private final SearchRequest request;
    }
```

?

Reply:
And then we can use that resolved pipeline for response transformation?

That could work, though it would mean that `SearchPipelineService` handles the request transformation, while the resolved `Pipeline` handles the response transformation.

Reply:
> And then we can use that resolved pipeline for response transformation?

That's one option, or delegate that to the `PipelinedSearchRequest` (fe, as `getResponse` method) as well (which could use `pipeline` for response transformation).

Reply:
I like that!

I'll get it updated shortly.